app/services/ClaimToAdjustService.scala

```diff
@@ -92,42 +98,37 @@ class ClaimToAdjustService @Inject()(val financialDetailsConnector: FinancialDet
   }
 
   private def checkCrystallisation(nino: Nino, taxYearList: List[TaxYear])(implicit hc: HeaderCarrier): Future[Option[TaxYear]] = {
-    taxYearList match {
-      case ::(head, next) => isTaxYearNonCrystallised(head, nino).flatMap {
-        case true => Future.successful(Some(head))
-        case false => checkCrystallisation(nino, next)
+    taxYearList.foldLeft(Future.successful(Option.empty[TaxYear])) { (acc, item) =>
+      acc.flatMap {
+        case Some(_) => acc
+        case None => isTaxYearNonCrystallised(item, nino) map {
+          case true => Some(item)
+          case false => None
+        }
       }
-      case Nil => Future.successful(None)
     }
   }
 
-  // Next 2 methods taken from Calculation List Service (with small changes), but I don't like the code duplication. Is there any solution to this?
-  private def isTaxYearNonCrystallised(taxYear: TaxYear, nino: Nino)(implicit hc: HeaderCarrier): Future[Boolean] = {
-
+  private def isFutureTaxYear(taxYear: TaxYear): Boolean = {
     val currentTaxYearEnd = dateService.getCurrentTaxYearEnd
-    val futureTaxYear = taxYear.endYear >= currentTaxYearEnd
-    if (futureTaxYear) {
+    taxYear.endYear >= currentTaxYearEnd
+  }
+
+  private def isTaxYearNonCrystallised(taxYear: TaxYear, nino: Nino)(implicit hc: HeaderCarrier): Future[Boolean] = {
+    if (isFutureTaxYear(taxYear)) {
       Future.successful(true)
     } else {
-      isTYSCrystallised(nino, taxYear.endYear).map {
-        case true => false
-        case false => true
+      calculationListConnector.getCalculationList(nino, taxYear.formatTaxYearRange).flatMap {
+        case res: CalculationListModel => Future.successful(res.crystallised)
+        case err: CalculationListErrorModel if err.code == 204 => Future.successful(Some(false))
+        case err: CalculationListErrorModel => Future.failed(new InternalServerException(err.message))
+      } map {
```

Comment:
you can change logic in lines 121-124 in the next way, so this code would be easy to read:

```
 calculationListConnector.getCalculationList(nino, taxYear.formatTaxYearRange).flatMap {
        case res: CalculationListModel => Future.successful(res.crystallised.getOrElse(false))
        case err: CalculationListErrorModel if err.code == 204 => Future.successful(false)
        case err: CalculationListErrorModel => Future.failed(new InternalServerException(err.message))
      }
```